File: Leetcode/Solutions/Classified-By-Company/Amazon Problems Solutions/535-encode-and-decode-tinyurl/535-encode-and-decode-tinyurl.java
public class Codec 
{
    String output = "http://tinyurl.com/";
    int[] count;
    int[] table;
    int hashcode = 0;
    // Encodes a URL to a shortened URL.
    public String encode(String longUrl)
    {
        int len = longUrl.length();
        count = new int[26];
        table = new int[len];
        for(int i = 0 ; i < len ; ++i)
        {
            table[i] = longUrl.charAt(i);
            hashcode += (i * (int)(longUrl.charAt(i)));
                if(Character.isLetter(longUrl.charAt(i)))
                    count[Character.toLowerCase(longUrl.charAt(i)) - 'a'] += 1;
        }
        for (int i = 0 ; i < 26 ; ++i)
            if(count[i] != 0)
            {
                output += (String.valueOf((char) (i + 'a') + "" + count[i]));
            }
        return output + " " + hashcode;
    }

    // Decodes a shortened URL to its original URL.
    public String decode(String shortUrl)
    {
        String encodedUrl = "";
        for(int i = 0 ; i < table.length ; ++i)
            encodedUrl += (char)(table[i]);
        return encodedUrl;
    }
}

// Your Codec object will be instantiated and called as such:
// Codec codec = new Codec();
// codec.decode(codec.encode(url));

File: Leetcode/Solutions/Classified-By-Company/Amazon Problems Solutions/535-encode-and-decode-tinyurl/535-encode-and-decode-tinyurl.cpp
class Solution {
public:
    unordered_map<string , string> urlDB , codeDB;
    // Encodes a URL to a shortened URL.
    
    string getCode()
    {
        string code = "http://tinyurl.com/";
        string chars = "1234567890abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ";
        for(int i = 0 ; i < 6 ; ++i)
            code += (chars[rand() % 62]);
        return code;
    }
    string encode(string longUrl) {
        if(urlDB.find(longUrl) != urlDB.end())
           return urlDB[longUrl];
        string code = getCode();
        while(codeDB.find(code) != codeDB.end())
            code = getCode();
        urlDB[longUrl] = code;
        codeDB[code] = longUrl;
        return code;
    }

    // Decodes a shortened URL to its original URL.
    string decode(string shortUrl) {
        return codeDB[shortUrl];
    }
};

// Your Solution object will be instantiated and called as such:
// Solution solution;
// solution.decode(solution.encode(url));